pkg/ipmi/handler.go

```diff
@@ -0,0 +1,45 @@
+// Copyright 2019-2022 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ipmi
+
+import (
+	"os"
+	"syscall"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+type syscalls interface {
+	syscall(uintptr, uintptr, uintptr, uintptr) (uintptr, uintptr, unix.Errno)
+	fileSyscallConn(f *os.File) (syscall.RawConn, error)
+	fileSetReadDeadline(f *os.File, t time.Duration) error
+	connRead(f func(fd uintptr) bool, conn syscall.RawConn) error
+}
+
+type realSyscalls struct{}
+
+func (r *realSyscalls) syscall(trap, a1, a2, a3 uintptr) (uintptr, uintptr, unix.Errno) {
+	return unix.Syscall(trap, a1, a2, a3)
+}
+
+func (r *realSyscalls) fileSyscallConn(f *os.File) (syscall.RawConn, error) {
+	return f.SyscallConn()
+}
+
+func (r *realSyscalls) fileSetReadDeadline(f *os.File, t time.Duration) error {
+	return f.SetReadDeadline(time.Now().Add(t))
+}
+
+func (r *realSyscalls) connRead(f func(fd uintptr) bool, conn syscall.RawConn) error {
+	return conn.Read(f)
+}
+
+type handler interface {
```

Comment:
qq: why is handler an interface? there seems to only be one implementation, the dev struct. I like that you broke it out into a separate struct, actually, it makes a lot of sense. but I think the interface could be removed and struct IPMI could just embed *dev?

Reply:
I am pretty sure we *might* need the interface in future, as more than just Linux systems will need ipmi interface (Plan 9 is getting it), so this is an attempt at future proofing it. Did I make a mistake in requesting it? I have no idea.

Reply:
ah I see. It's one of those things that can be borderline.

I tend to /not/ do interfaces in these cases, because at compile time we will choose one GOOS or another -- which means at compile time there is still only one implementation of handler. And the Plan 9 version could also be called dev. Interfaces are good for cases where at run time there is more than one implementation choice for a given thing. i don't feel  too strongly about it though.

Reply:
So, what is the decision here? Should I rework it or do we keep it the way it is?

Reply:
Here is where this gets weird and why it's hard for me to say. 

Will this test run on a plan 9 system, and use an imported /sys from a linux system? Or vice versa? That's extremely unlikely, but  my world, not uncommon. But at that point, we might someday choose to support multiple GOOS in one binary ... but that's a very unlikely situation?

Anyway, I tend to trust Chris K on stuff like this, so I'd take his direction; or, since he does not feel too strongly, this may be a case where we should let you decide, since the big goal here is to have 9e make more decisions of this type as time goes on. Are you comfortable making this decision? I am fine with whatever you wish to do.

As you can see, it's kind of borderline.

Reply:
Ok, I deleted the interface, also the mock and the related tests because they did not contribute any additional value except coverage of execution but not correctness. The most stuff is tested in QEMU anyway.

Reply:
Oh! Now I remember why I had the mock and the interface. I had to prevent using the dev.SendRequest method, because it calls f.RawConn, which is not possible with a normal file. The mock allowed me to get coverage for the most top level ipmi functions. I couldnt figure out how to create a file with the required properties to satisfy the call to SyscallConn.
It was an additional layer of abstraction to actually run the functions without an ipmi interface available. 
Now we run the tests with the ipmi interface available, but the device does not support all features implemented. So we can't cover all functions. The problem with the file rawconn still persists in this environment, because as soon as I try to abstract the interface with a normal file it wont work.